### PR TITLE
package/plumed: added fine grained control over optional modules

### PR DESCRIPTION
**Pull request by marcodelapierre**
In this PR, I am editing the package recipe for PLUMED. 
In particular, I am adding a more fine grained control for enabling optional modules.
I have used a similar syntax to the LAMMPS recipe and its packages.

As a result:
* the existing variant `optional_modules` now only allows to set values of `all`, `reset`, or `none`.
* each module has its own variant: use `+<module>` to enable its installation
* note that the set of available modules changes over PLUMED versions; conflicts are managed appropriately
* as a result of this implementation, each individual module variant only works to activate the module, not to de-activate it
* specific modules can only be activated in conjunction with `optional_modules` equal to `none`, or `reset`, whereas they conflict with `all` (because the latter already has all modules installed, by definition)
 
[EDIT]: I have also added myself as maintainer, see chat in https://github.com/plumed/plumed2/issues/719

### CODE REVIEW

**Comment by tldahlgren** [comment]
@tgamblin @w8jcik @RemiLacroix-IDRIS As contributors of more than a few lines to this package, are you any of you interested in commenting on this PR?

**Comment by tldahlgren** [comment]
@spackbot run pipeline

**Comment by RemiLacroix-IDRIS** [comment]
I understand the rational but it feels a bit weird that you have `optional_modules=all` and `~<module>` since it actually means `+<module>`.

Not sure if there is a better solution though so let see if others have some opinions.

File: var/spack/repos/builtin/packages/plumed/package.py [hunk 1]
```diff
@@ -87,6 +116,29 @@ class Plumed(AutotoolsPackage):
...
+    conflicts('+ves', when='@:2.3.99', msg='ves was added from version 2.4')
+    conflicts('+dimred', when='@:2.4.99', msg='dimred was added from version 2.5')
+    conflicts('+logmfd', when='@:2.4.99', msg='logmfd was added from version 2.5')
+    conflicts('+pamm', when='@:2.4.99', msg='pamm was added from version 2.5')
+    conflicts('+piv', when='@:2.4.99', msg='piv was added from version 2.5')
+    conflicts('+annfunc', when='@:2.5.99', msg='annfunc was added from version 2.6')
+    conflicts('+maze', when='@:2.5.99', msg='maze was added from version 2.6')
+    conflicts('+fisst', when='@:2.6.99', msg='fisst was added from version 2.7')
+    conflicts('+funnel', when='@:2.6.99', msg='funnel was added from version 2.7')
+    conflicts('+opes', when='@:2.6.99', msg='opes was added from version 2.7')
```
Comment (alalazo) [inline var/spack/repos/builtin/packages/plumed/package.py]: This can be rewritten using #29530 and a multi-valued variant. You can check `nsimd` for an example of that.

**Review by alalazo — CHANGES_REQUESTED** [review]
Can you check if the new feature I linked can help addressing the comments in the recipe and @RemiLacroix-IDRIS concerns?

**Comment by marcodelapierre** [comment]
Thanks @RemiLacroix-IDRIS and @alalazo for the feedback and patience.
I have addressed your requests today, by re-implementing the variant `optional_modules` using the multi-valued variant approach (`disjoint_sets`) in conjunction with the new `conditional()` functionality to clean up conflicts.

I have also taken the chance to add 3 new modules that come with the latest version `2.8` of the package.

Done a few concretisation and install tests with some meaningful combination of values for `optional_modules`, seems good to go to me.

**Comment by GiovanniBussi** [comment]
@marcodelapierre thanks for taking care of this!

I was double checking and I realized that modules `reference` and `vesselbase` cannot be switched off (they are `always` on, see files `reference/module.type` and `vesselbase/module.type`. Thus, configuring with
`./configure --enable-modules=-reference` has no effects. Notice that this should report an error during `./configure` ([it doesn't](https://github.com/plumed/plumed2/issues/833)).

Anyway, I was checking if the decision of keeping `reference` and `vesselbase` as `always` is motivated.

These modules are used in many other modules:
```
reference: analysis colvar dimred function mapping secondarystructure
vesselbase: adjmat analysis crystallization gridtools manyrestraints mapping multicolvar secondarystructure
```

For comparison, here is where the `default-on` modules are used:
```
analysis: dimred vesselbase
bias: analysis drr eds fisst funnel isdb logmfd maze opes ves vesselbase
colvar: funnel isdb maze membranefusion piv s2cm
function: annfunc isdb pytorch
molfile: cltools
multicolvar: adjmat analysis crystallization pamm
xdrfile: cltools generic
```
and here the same analysis for `default-off` modules:
```
adjmat: pamm
```

I would say that if we keep the `bias` module as `default-on`, also `reference` and `vesselbase` should be `default-on`. I will apply this to master branch in plumed not to change the behavior of published releases.

**Comment by RemiLacroix-IDRIS** [comment]
Looks good to me, keeping in mind @GiovanniBussi's comment and any possible implications it might have.

File: var/spack/repos/builtin/packages/plumed/package.py (hunk 1 above)
Comment (marcodelapierre) [reply var/spack/repos/builtin/packages/plumed/package.py]: This now seems resolved to me @alalazo

**Comment by marcodelapierre** [comment]
Hi @GiovanniBussi , my pleasure.

What would you think if we were to go ahead with this PR, and then add any additional conflict checks / logics to individual modules in future PRs?
Or do you think that, amongst your comments above, there are some crucial fixes that need to be applied straight away?

I will further comment on this topic in the Plumed issue you mentioned above. Happy to chat further there, to further improve Plumed support in Spack.

**Comment by GiovanniBussi** [comment]
@marcodelapierre I think this PR is fine as it is, the error is on our side.

In addition, notice that there are some dependencies between modules. This implies that some combinations will **not** work. However, I am not sure it is worth to add the logic checking for these dependencies in spack.

Honestly, I am not sure even if the `none` option is necessary. I think the pratical use cases are:
- normal build (everything that is on by default)
- one or more optional modules is switched on

I cannot imagine a final user willing to build plumed without ANY collective variable. This was mostly added to allow compiling in less than a minute a stripped down version of PLUMED that can be used to test tools.

**Comment by marcodelapierre** [comment]
Ok, makes sense, thanks @GiovanniBussi .
I will make an extra commit soon (and test it) to  get rid of the none option, and update the variant description.

**Comment by marcodelapierre** [comment]
Done!

**Comment by marcodelapierre** [comment]
Final add done: added dependency of `pamm` on `adjmat`, within the variant `optional_modules`.
(as per @GiovanniBussi 's suggestion here: https://github.com/plumed/plumed2/pull/834#issuecomment-1170867805)


I won't do any further commits, I think it is ready for final review, @alalazo , @RemiLacroix-IDRIS

File: var/spack/repos/builtin/packages/plumed/package.py [hunk 2]
```diff
@@ -68,13 +69,61 @@ class Plumed(AutotoolsPackage):
...
+                               conditional('imd', when='@:2.2.99'),
+                               conditional('isdb', when='@2.4:'),
+                               conditional('logmfd', when='@2.5:'),
+                               'manyrestraints', 'mapping',
+                               conditional('maze', when='@2.6:'),
+                               'molfile',
+                               'multicolvar',
+                               conditional('opes', when='@2.7:'),
+                               conditional('pamm',
+                               when='@2.5: optional_modules=adjmat'),
```
Comment (RemiLacroix-IDRIS) [inline var/spack/repos/builtin/packages/plumed/package.py]: That's nitpicking but I would align `when=...` with `'pamm'.

Also is the concretizer really clever enough to deal with this (I mean you have a multivalued variant with a value depending on another value of that same variant)? If it is then it's great!

**Review by RemiLacroix-IDRIS — CHANGES_REQUESTED** [review]
Aside from the style comment and concretizer question about "pamm", LGTM.

**Comment by marcodelapierre** [comment]
> Also is the concretizer really clever enough to deal with this (I mean you have a multivalued variant with a value depending on another value of that same variant)? If it is then it's great!

Very good question @RemiLacroix-IDRIS , on the mutual dependency between two values that are part of a multi-valued variant set.
I did test it, and the answer is: YES, the concretiser is quite smart!
What it does is to automatically add  the required value to the variant, e.g. you ask for `optional_modules=pamm`, and it turns it into `optional_modules=adjmat,pamm`.
I guess it can do it because it is a multi-valued variants, and it knows the corresponding logic of this construct.

File: var/spack/repos/builtin/packages/plumed/package.py (hunk 2 above)
Comment (marcodelapierre) [reply var/spack/repos/builtin/packages/plumed/package.py]: On the concretiser I have replied in the main thread, because I think it is a very important point!

On the alignment, I did break it into lines, because the single line would be 89 characters wide, and I think the Flake8 style check would bite me (but I may be wrong).

Comment (RemiLacroix-IDRIS) [reply var/spack/repos/builtin/packages/plumed/package.py]: I meant:
```python
                               'multicolvar',
                               conditional('opes', when='@2.7:'),
                               conditional('pamm',
                                           when='@2.5: optional_modules=adjmat'),
                               conditional('piv', when='@2.5:'),
                               conditional('reference', when='@:2.2.99'),
```
so that's it's clearer that `when` applies to that specific `conditional` construct.

Comment (marcodelapierre) [reply var/spack/repos/builtin/packages/plumed/package.py]: Oh I see!
Just done it now with a new commit.

**Comment by RemiLacroix-IDRIS** [comment]
>  I did test it, and the answer is: YES, the concretiser is quite smart!

Good, @alalazo will know anyway.

**Comment by alalazo** [comment]
@RemiLacroix-IDRIS The concretizer does a global solve, so yes it can deal with these kind of things :slightly_smiling_face:

**Review by alalazo — APPROVED** [review]
@RemiLacroix-IDRIS Can you approve the PR if you have no further requests?

**Review by RemiLacroix-IDRIS — APPROVED** [review]
Yup, all good for me!
